bootloader/efi/efi.go

```diff
@@ -0,0 +1,193 @@
+const (
+	VariableNonVolatile       VariableAttr = 0x00000001
+	VariableBootServiceAccess VariableAttr = 0x00000002
+	VariableRuntimeAccess     VariableAttr = 0x00000004
+)
+
+var (
+	isSnapdTest = len(os.Args) > 0 && strings.HasSuffix(os.Args[0], ".test")
+	openEFIVar  = openEFIVarImpl
+)
+
+const expectedEFIvarfsDir = "/sys/firmware/efi/efivars"
+
+func openEFIVarImpl(name string) (r io.ReadCloser, attr VariableAttr, size int64, err error) {
+	mounts, err := osutil.LoadMountInfo()
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	found := false
+	for _, mnt := range mounts {
+		if mnt.MountDir == expectedEFIvarfsDir {
+			if mnt.FsType == "efivarfs" {
+				found = true
+				break
+			}
+		}
+	}
+	if !found {
+		return nil, 0, 0, ErrNoEFISystem
+	}
+	varf, err := os.Open(filepath.Join(dirs.GlobalRootDir, expectedEFIvarfsDir, name))
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	defer func() {
+		if err != nil {
+			varf.Close()
+		}
+	}()
+	fi, err := varf.Stat()
+	if err != nil {
+		return nil, 0, 0, err
+	}
+	sz := fi.Size()
+	if sz < 4 {
+		return nil, 0, 0, fmt.Errorf("unexpected size: %d", sz)
+	}
+
+	var abuf [4]byte
+	if _, err = varf.Read(abuf[:]); err != nil {
+		return nil, 0, 0, err
+	}
+	return varf, VariableAttr(binary.LittleEndian.Uint32(abuf[:])), sz - 4, nil
+}
+
+func cannotReadError(name string, err error) error {
+	return fmt.Errorf("cannot read EFI var %q: %v", name, err)
+}
+
+// ReadVarBytes will attempt to read the bytes of the value of the
+// specified EFI variable, specified by its full name composed of the
+// variable name and vendor ID. It also returns the attribute value
+// attached to it. It expects to use the efivars filesystem at
+// /sys/firmware/efi/efivars.
+// https://www.kernel.org/doc/Documentation/filesystems/efivarfs.txt
+// for more details.
+func ReadVarBytes(name string) ([]byte, VariableAttr, error) {
+	varf, attr, _, err := openEFIVar(name)
+	if err != nil {
+		if err == ErrNoEFISystem {
+			return nil, 0, err
+		}
+		return nil, 0, cannotReadError(name, err)
+	}
+	defer varf.Close()
+	b, err := ioutil.ReadAll(varf)
+	if err != nil {
+		return nil, 0, cannotReadError(name, err)
+	}
+	return b, attr, nil
+}
+
+// ReadVarString will attempt to read the string value of the
+// specified EFI variable, specified by its full name composed of the
+// variable name and vendor ID. The string value is expected to be
+// encoded as UTF16. It also returns the attribute value attached to
+// it. It expects to use the efivars filesystem at
+// /sys/firmware/efi/efivars.
+// https://www.kernel.org/doc/Documentation/filesystems/efivarfs.txt
+// for more details.
+func ReadVarString(name string) (string, VariableAttr, error) {
+	varf, attr, sz, err := openEFIVar(name)
+	if err != nil {
+		if err == ErrNoEFISystem {
+			return "", 0, err
+		}
+		return "", 0, cannotReadError(name, err)
+	}
+	defer varf.Close()
+	if sz%2 != 0 {
+		return "", 0, fmt.Errorf("EFI var %q has an extra byte to be an UTF16 string", name)
```

Comment:
this reads awkwardly, maybe
```suggestion
		return "", 0, fmt.Errorf("EFI var %q isn't a valid UTF16 string, it has an extra byte", name)
```